### PR TITLE
nixos/mastodon: update redis host and port configurations

### PR DESCRIPTION
**Pull request by Izorkin**
## Description of changes
Update redis host and port configurations and optimize systemd startup order block.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/mastodon.nix [hunk 1]
```diff
@@ -90,6 +90,10 @@ let
     SystemCallArchitectures = "native";
   };
 
+  commonServices = lib.optional redisActuallyCreateLocally "redis-mastodon.service"
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:93]:
```suggestion
  # Services that all Mastodon units After= and Requires= on
  commonServices = lib.optional redisActuallyCreateLocally "redis-mastodon.service"
```

Can we have an explaining comment for this?

Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:93]: Ok, update PR.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 2]
```diff
@@ -631,6 +630,13 @@ in {
 
   config = lib.mkIf cfg.enable (lib.mkMerge [{
     assertions = [
+      {
+        assertion = !redisActuallyCreateLocally -> (cfg.redis.host != "127.0.0.1" && cfg.redis.port != null);
+        message = ''
+          <option>services.mastodon.redis.host</option> and <option>services.mastodon.redis.port</option>
+            needs to be set if <option>services.mastodon.redis.createLocally</option> is not enabled.
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:637]:
```suggestion
          `services.mastodon.redis.host` and `services.mastodon.redis.port` need to be set if `services.mastodon.redis.createLocally` is not enabled.
```

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/web-apps/mastodon.nix (hunk 2 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix]: Fixed, thanks!

File: nixos/modules/services/web-apps/mastodon.nix [hunk 3]
```diff
@@ -34,6 +32,8 @@ let
 
     TRUSTED_PROXY_IP = cfg.trustedProxy;
   }
+  // lib.optionalAttrs (cfg.redis.host != null && !cfg.redis.enableUnixSocket) { DB_HOST = cfg.database.host; }
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:35]:
```suggestion
  // lib.optionalAttrs (cfg.redis.host != null && !cfg.redis.enableUnixSocket) { REDIS_HOST = cfg.redis.host; }
```
Is this a typo?

**Review by erictapen — CHANGES_REQUESTED** [review]
One last bit, I think

File: nixos/modules/services/web-apps/mastodon.nix (hunk 3 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix]: Oops, fixed. Thanks!

**Review by erictapen — APPROVED** [review]

**Comment by erictapen** [comment]
@GrahamcOfBorg test mastodon

**Review by erictapen — CHANGES_REQUESTED** [review]
This breaks `nixosTests.mastodon.remote-databases` ...

```
vm-test-run-mastodon-remote-postgresql> server # [  101.929081] run-streaming.sh[1351]: ERR! Redis Client Error! Error: connect ECONNREFUSED 127.0.0.1:6379
vm-test-run-mastodon-remote-postgresql> server # [  101.930220] run-streaming.sh[1351]: ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1605:16)
vm-test-run-mastodon-remote-postgresql> server # [  101.931357] run-streaming.sh[1351]: ERR! Redis Client Error!  Error: connect ECONNREFUSED 127.0.0.1:6379
vm-test-run-mastodon-remote-postgresql> server # [  101.931958] run-streaming.sh[1351]: ERR! Redis Client Error!     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1605:16) {
vm-test-run-mastodon-remote-postgresql> server # [  101.932731] run-streaming.sh[1351]: ERR! Redis Client Error!   errno: -111,
vm-test-run-mastodon-remote-postgresql> server # [  101.933192] run-streaming.sh[1351]: ERR! Redis Client Error!   code: 'ECONNREFUSED',
vm-test-run-mastodon-remote-postgresql> server # [  101.933902] run-streaming.sh[1351]: ERR! Redis Client Error!   syscall: 'connect',
vm-test-run-mastodon-remote-postgresql> server # [  101.934586] run-streaming.sh[1351]: ERR! Redis Client Error!   address: '127.0.0.1',
vm-test-run-mastodon-remote-postgresql> server # [  101.935212] run-streaming.sh[1351]: ERR! Redis Client Error!   port: 6379
vm-test-run-mastodon-remote-postgresql> server # [  101.935838] run-streaming.sh[1351]: ERR! Redis Client Error! }
```

**Comment by erictapen** [comment]
Now the standard test fails at eval...

Can you please check for the tests to work before pushing? Makes review easier.

```
$ nix build .#nixosTests.mastodon.standard -L
error:
       … while evaluating a branch condition

         at /nix/store/wqil6dm92m92x2ksqkd6a2m8q7hml6l2-source/nixos/tests/all-tests.nix:21:5:

           20|   discoverTests = val:
           21|     if isAttrs val
             |     ^
           22|     then

       … while calling the 'isAttrs' builtin

         at /nix/store/wqil6dm92m92x2ksqkd6a2m8q7hml6l2-source/nixos/tests/all-tests.nix:21:8:

           20|   discoverTests = val:
           21|     if isAttrs val
             |        ^
           22|     then

       (stack trace truncated; use '--show-trace' to show the full trace)

       error:
       Failed assertions:
       - `services.mastodon.redis.enableUnixSocket` needs to be disabled if
         `services.mastodon.redis.host` and `services.mastodon.redis.port` is used.
```

**Comment by Izorkin** [comment]
Sorry, I forgot to check the previous fix.
Update PR.

**Review by erictapen — APPROVED** [review]
Now both tests build for me on x86_64. Thanks!

**Comment by Izorkin** [comment]
Thanks!

**Pull request merged**